src/tools/cgclassify.c

```diff
@@ -198,7 +231,459 @@ int main(int argc, char *argv[])
+		return -1;
+	}
+
+	snprintf(cgroup_procs_path, sizeof(cgroup_procs_path), "%s/%s/cgroup.procs",
+		 cgroup_name, scope_name);
+	free(cgroup_name);
+
+	cgroup_systemd_path_f = fopen(cgroup_procs_path, "we");
+	if (!cgroup_systemd_path_f) {
+		err("Failed to open %s\n", cgroup_procs_path);
+		return -1;
+	}
+
+	if (is_cgroup_mode_hybrid()) {
+		/*
+		 * construct the unified cgroup path, by parsing
+		 * /proc/mounts
+		 */
+		find_mnt_point("unified", &cgroup_name);
+		if (!cgroup_name) {
+			err("Unable find unified cgroup path\n");
+			fclose(cgroup_systemd_path_f);
+			return -1;
+		}
+
+		snprintf(cgroup_procs_path, sizeof(cgroup_procs_path), "%s/%s/cgroup.procs",
+			 cgroup_name, scope_name);
+		free(cgroup_name);
+
+		cgroup_unified_path_f = fopen(cgroup_procs_path, "we");
+		if (!cgroup_unified_path_f) {
+			err("Failed to open %s\n", cgroup_procs_path);
+			fclose(cgroup_systemd_path_f);
+			return -1;
+		}
+	}
+
+	fprintf(cgroup_systemd_path_f, "%d", pid);
+	fflush(cgroup_systemd_path_f);
+	fclose(cgroup_systemd_path_f);
+
+	if (!is_cgroup_mode_hybrid())
+		return 0;
+
+	fprintf(cgroup_unified_path_f, "%d", pid);
+	fflush(cgroup_unified_path_f);
+	fclose(cgroup_unified_path_f);
+
+	return 0;
+}
+
+static int is_scope_parsed(const char * const path)
+{
+	/*
+	 * As per, <kernel sources>/kernel/cgroup/cgroup.c::cgroup_init()
+	 * At the max there can be only 16 controllers and we are
+	 * not accounting for named hierarchies, which can be more
+	 * than 16 themselves.
+	 */
+	static char parsed_scope_path[MAX_MNT_ELEMENTS][FILENAME_MAX];
+	int i;
+
+	for (i = 0; i < MAX_MNT_ELEMENTS; i++) {
+		if (!strcmp(parsed_scope_path[i], path))
+			return 1;
+		if (parsed_scope_path[i][0] == '\0') {
+			snprintf(parsed_scope_path[i], FILENAME_MAX, "%s", path);
+			break;
+		}
+	}
+
+	return 0;
+}
+
+/* Borrowed from src/api.c::__attach_task_pid */
+static int attach_task_pid(char *path, pid_t tid)
+{
+	FILE *tasks = NULL;
+	int ret = 0;
+
+	tasks = fopen(path, "we");
+	if (!tasks) {
+		switch (errno) {
+		case EPERM:
+			ret = ECGROUPNOTOWNER;
+			break;
+		case ENOENT:
+			ret = ECGROUPNOTEXIST;
+			break;
+		default:
+			ret = ECGROUPNOTALLOWED;
+		}
+		goto err;
+	}
+	ret = fprintf(tasks, "%d", tid);
+	if (ret < 0) {
+		ret = ECGOTHER;
+		goto err;
+	}
+	ret = fflush(tasks);
+	if (ret) {
+		ret = ECGOTHER;
+		goto err;
+	}
+	fclose(tasks);
+	return 0;
+err:
+	err("cannot write tid %d to %s:%s\n", tid, path, strerror(errno));
+	if (tasks)
+		fclose(tasks);
+	return ret;
+}
+
+static int rollback_pid_cgroups(pid_t pid)
+{
+	char cgroup_proc_path[FILENAME_MAX + 14];
+	char cgroup_path[FILENAME_MAX];
+	int err = 0, idx = 0, ret = 0;
+	char *cgrp_proc_path = NULL;
+
+	/*
+	 * unified cgroup rollback is simple, we need to write into
+	 * single cgroup hierarchy.
+	 */
+	if (is_cgroup_mode_unified()) {
+		pthread_rwlock_rdlock(&cg_mount_table_lock);
+		cg_build_path_locked(info[idx].cgrp_path, cgroup_path, NULL);
```

Comment:
What if a user does something like this:
```
cgcreate -c -S -g cpu:my.slice/db.scope   # set my.slice/db.scope as the default
cgclassify -r -g cpu:my.slice/db.scope 1234
```

This line would have `cg_build_path_locked()` prepend `my.slice/db.scope` in front of `info[idx].cgroup_path`.  But since `info[idx].cgroup_path` came directly from `/proc/{pid}/cgroup`, I don't see why we would ever want to prepend the slice/scope info.  What am I missing?

Reply:
In this above example the user is expected to use one of the following commands:
`cgclassify -b -r -g cpu:my.slice/db.scope 1234` or
`cgclassify -r -g cpu:/ 1234`

either the user could use `-b` option to ignore the default delegated `slice/scope` or pass `/` referring to the root a.k.a delegated `slice and scope`

Reply:
But since the cgroup path info is coming from `/proc/{pid}/cgroup`, the slice/scope are already there right?  So, wouldn't your first example end up with a path like `/sys/fs/cgroup/my.slice/db.scope/my.slice/db.scope`?

Reply:
Gotcha.  I missed the `-b` in your first example. Thanks!